feat(*): add SDK core (#1)
deepsourcelabs/deepsource-go: guides/writing-analyzers.md

The flow of writing a custom analyzer using the SDK, is as follows:
- Create an analyzer (`CLIRunner`)
- Run the analyzer (`Run()`)
- Use the processor to fetch a DeepSource-compatible report (`Processor.Process()`)
- Persist the report to the local filesystem using `SaveReport`

## Getting Started

### Setting up our analyzer

The analyzer should contain the following:
- `Name`: Name of the analyzer
- `Command`: The main command for the CLI tool (for example, `staticcheck`, etc.)
- `Args`: Arguments for `Command`.
- `Processor`: Processor used for parsing the output of the CLI analyzer

The analyzer can be executed using `Run()`, which executes the CLI (`Command` along with its `Args`). The output of the CLI is stored to its respective buffers. (`stdout` and `stderr`; accessible through `a.Stdout()` and `a.Stderr()`)

Here is the code for the analyzer:

```go
package main

import (
	"fmt"
	"log"

	"github.com/deepsourcelabs/deepsource-go/analyzers"
	"github.com/deepsourcelabs/deepsource-go/analyzers/processors"
)

func main() {
	a := analyzers.CLIRunner{
		Name:      "staticcheck",
		Command:   "staticcheck",
		Args:      []string{"-f", "text", "./..."},

> 		Processor: &processor, <=== will be implemented later

Reviewer: ```suggestion
		Processor: &processor, // <=== will be implemented later
```

						"line": 5,
						"column": 6
					},
					"end": {
						"line": 0,
						"column": 0
					}
				}
			}
		},
		{
			"issue_code": "SA4017",
			"issue_text": "Sprint is a pure function but its return value is ignored ",
			"location": {
				"path": "analyzers/testdata/src/staticcheck/staticcheck.go",
				"position": {
					"begin": {
						"line": 6,
						"column": 2
					},
					"end": {
						"line": 0,
						"column": 0
					}
				}
			}
		},
		{
			"issue_code": "S1039",
			"issue_text": "unnecessary use of fmt.Sprint ",
			"location": {
				"path": "analyzers/testdata/src/staticcheck/staticcheck.go",
				"position": {
					"begin": {
						"line": 6,
						"column": 2
					},
					"end": {
						"line": 0,
						"column": 0
					}
				}
			}
		}
	],
	"errors": null,
	"extra_data": null
}
```

## Generating TOML files for issues

The developer can define all issues in a single TOML file. This file acts as a single point of truth for generating TOML files for each issue.

This is helpful for developers who wish to define custom issues for their analyzers.

For example, we have `issues.toml` as the file containing details for all issues:

```toml
[[issues]]

> IssueCode = "SA4017"
> Category = "bug-risk"
> Title = "Sprint is a pure function but its return value is ignored"
> Description = """

Reviewer: ```suggestion
issue_code = "SA4017"
category = "bug-risk"
title = "Sprint is a pure function but its return value is ignored"
description = """
Pure functions do not change the passed value but return a new value that is meant to be used. This issue is raised when values returned by pure functions are discarded.
```

> ## Sample

Reviewer: ```suggestion
```

"""
```

> **Note**:
>
> `GenerateTOML` requires `REPO_ROOT` to be set in the environment variables.

`GenerateTOML` reads `$REPO_ROOT/.deepsource/analyzers/issues.toml`, and generates TOML files for each issue, where the filename is the issue code.

The TOML files are generated at `$REPO_ROOT/.deepsource/analyzers/issues/<IssueCode>.toml`.

```go
    // previous code

    // generate TOML files for each issue from a parent TOML file
	err = GenerateTOML()
	if err != nil {
		log.Fatalln(err)
	}
```

On inspecting `$REPO_ROOT/.deepsource/analyzers/issues/SA4017.toml`, we can see the following contents:

```toml
IssueCode = "SA4017"
Category = "bug-risk"
Title = "Sprint is a pure function but its return value is ignored"

> Description = "<h2>Sample</h2>\n"

Reviewer: ```suggestion
issue_code = "SA4017"
category = "bug-risk"
title = "Sprint is a pure function but its return value is ignored"
description = "Pure functions do not change the passed value but return a new value that is meant to be used. This issue is raised when values returned by pure functions are discarded."
```

> Code = "SA4017"
> Text = "Sprint is a pure function but its return value is ignored"
> ShortDescription = "Sprint is a pure function but its return value is ignored"
> Description = "<h2>Sample</h2>\n"

Reviewer: ```suggestion
code = "SA4017"
title = "Sprint is a pure function but its return value is ignored"
category = "bug-risk"
description = """Sample description"""
```

> ```

Reviewer: Also add docs on how devs can write tests for their analyzers.
